# AI TRAINING – WEEK 2
## DAY 5 ASSIGNMENT: NumPy Arrays, Dimensions, Indexing, Slicing & Vectorized Operations
**Learn → Create Arrays → Access Data → Process Efficiently**

- **Student Name:** Lasya Kolluru
- **GitHub Username:** Lasya-Kolluru
- **Email:** kollurlasyachoudri@gmail.com
- **Deliverables:** `numpy_assignment_Lasya_Kolluru.py`, `Week2_Day5_NumPy_Basics_Lasya_Kolluru.ipynb`
- **Core Topics:** 1D/2D/3D Array Creation, `ndim` vs `shape` vs `size`, Multi-Axis Indexing, Slicing Subarrays, Vectorized Arithmetic & Boolean Masking, NumPy Debugging Labs, Student Marks Analyzer Project, and Theoretical Reflections.



---
## Part A – Create NumPy Arrays
NumPy arrays (`ndarray`) are homogeneous, contiguous multidimensional containers that offer tremendous performance advantages over built-in Python lists.

In this section:
1. Create a 1D array from `[10, 20, 30, 40, 50]`.
2. Create a 2D array from `[[10, 20, 30], [40, 50, 60]]`.
3. Create a 3D array using values 1 to 8, arranged into two 2x2 blocks.
4. Inspect array representations and their `dtype`.
5. Create arrays of zeros and ones using `np.zeros()` and `np.ones()`.



In [1]:
import numpy as np

# 1. 1D Array
arr1 = np.array([10, 20, 30, 40, 50])

# 2. 2D Array
arr2 = np.array([[10, 20, 30],
                 [40, 50, 60]])

# 3. 3D Array
arr3 = np.array([[[1, 2],
                  [3, 4]],
                 [[5, 6],
                  [7, 8]]])

# 4. Print arrays and dtypes
print("1D Array:")
print(arr1)
print("Data type:", arr1.dtype)

print("\n2D Array:")
print(arr2)
print("Data type:", arr2.dtype)

print("\n3D Array:")
print(arr3)
print("Data type:", arr3.dtype)

# 5. Zeros and Ones
zeros = np.zeros(5)
ones = np.ones(5)

print("\nArray of zeros:")
print(zeros)
print("Zeros dtype:", zeros.dtype)

print("\nArray of ones:")
print(ones)
print("Ones dtype:", ones.dtype)


1D Array:
[10 20 30 40 50]
Data type: int32

2D Array:
[[10 20 30]
 [40 50 60]]
Data type: int32

3D Array:
[[[1 2]
  [3 4]]
 [[5 6]
  [7 8]]]
Data type: int32

Array of zeros:
[0. 0. 0. 0. 0.]
Zeros dtype: float64

Array of ones:
[1. 1. 1. 1. 1.]
Ones dtype: float64


### Key Observations:
- **Homogeneous Data**: All elements within a standard NumPy array share an identical data type (`int32`/`int64` for integers, `float64` for floats).
- **Default Floats in Builders**: Built-in factory functions like `np.zeros()` and `np.ones()` default to `float64` precision.



---
## Part B – Dimensions and Shape
Understanding the geometric layout of an array is vital in data engineering and machine learning.

- `ndim`: Number of axes (dimensions).
- `shape`: A tuple indicating the size along each axis.
- `size`: The total number of elements contained in the array.



In [2]:
print("1D Array")
print("ndim:", arr1.ndim)
print("shape:", arr1.shape)
print("size:", arr1.size)

print("\n2D Array")
print("ndim:", arr2.ndim)
print("shape:", arr2.shape)
print("size:", arr2.size)

print("\n3D Array")
print("ndim:", arr3.ndim)
print("shape:", arr3.shape)
print("size:", arr3.size)

print("\nZeros Array")
print("ndim:", zeros.ndim)
print("shape:", zeros.shape)
print("size:", zeros.size)

print("\nOnes Array")
print("ndim:", ones.ndim)
print("shape:", ones.shape)
print("size:", ones.size)

# Predict shape of 3-row, 4-column array
print("\nPredicted shape for a 3-row, 4-column array:", (3, 4))


1D Array
ndim: 1
shape: (5,)
size: 5

2D Array
ndim: 2
shape: (2, 3)
size: 6

3D Array
ndim: 3
shape: (2, 2, 2)
size: 8

Zeros Array
ndim: 1
shape: (5,)
size: 5

Ones Array
ndim: 1
shape: (5,)
size: 5

Predicted shape for a 3-row, 4-column array: (3, 4)


### Comparison Table: ndim, shape, and size

| Attribute | What It Represents | Example (2D Array) | Mathematical Formula |
| :--- | :--- | :--- | :--- |
| `ndim` | Number of axes (dimensions) | `2` | `len(array.shape)` |
| `shape` | Length of each dimension | `(2, 3)` | (rows, columns) |
| `size` | Total element count | `6` | rows * columns |



---
## Part C – Indexing
NumPy indexing provides fast direct-element access and in-place mutability.

1. Positive indexing from `0` to `n-1`.
2. Negative indexing from `-1` backwards.
3. In-place mutation via assignment.
4. Multi-dimensional tuple-like indexing: `arr[row_index, col_index]`.



In [3]:
marks = np.array([72, 85, 91, 66, 78, 95])

print("First element:", marks[0])
print("Last element:", marks[-1])
print("Element at index 2:", marks[2])
print("Second-last element:", marks[-2])

# Modify element at index 3
marks[3] = 70
print("Updated array:", marks)

# 2D Array access
arr = np.array([[11, 12, 13],
                [21, 22, 23]])

print("\nElement 22 (row 1, col 1):", arr[1, 1])
print("Element 13 (row 0, col 2):", arr[0, 2])


First element: 72
Last element: 95
Element at index 2: 91
Second-last element: 78
Updated array: [72 85 91 70 78 95]

Element 22 (row 1, col 1): 22
Element 13 (row 0, col 2): 13


---
## Part D – Slicing
Slicing enables extracting subsets of data using the syntax `[start:stop:step]`.

- Slices create **views** rather than copies where possible.
- The `stop` index is non-inclusive (`[start, stop)` half-open interval).
- Multi-dimensional arrays use comma-separated slice definitions: `arr[row_slice, col_slice]`.



In [4]:
values = np.array([5, 10, 15, 20, 25, 30, 35, 40])

print("First four elements:", values[:4])
print("Index 2 to 5:", values[2:6])
print("Every second element:", values[::2])
print("Last three elements:", values[-3:])

matrix = np.array([[1, 2, 3],
                   [4, 5, 6],
                   [7, 8, 9]])

print("\nFirst two rows and last two columns:\n", matrix[:2, -2:])

print("\nRationale:")
print("The stop index is excluded because Python slicing adheres to the half-open interval [start, stop).")
print("This guarantees that the length of the slice is strictly (stop - start),")
print("and consecutive slices arr[:k] and arr[k:] partition the array without gaps or overlaps.")


First four elements: [ 5 10 15 20]
Index 2 to 5: [15 20 25 30]
Every second element: [ 5 15 25 35]
Last three elements: [30 35 40]

First two rows and last two columns:
[[2 3]
 [5 6]]

Rationale:
The stop index is excluded because Python slicing adheres to the half-open interval [start, stop).
This guarantees that the length of the slice is strictly (stop - start),
and consecutive slices arr[:k] and arr[k:] partition the array without gaps or overlaps.


---
## Part E – Vectorized Operations
Vectorization is the hallmark of high-performance scientific computing in Python.
Instead of looping element-by-element at the slow Python bytecode level, operations are delegated to pre-compiled C routines and SIMD (Single Instruction, Multiple Data) CPU instructions.



In [5]:
a = np.array([10, 20, 30, 40])
b = np.array([1, 2, 3, 4])

print("Addition:", a + b)
print("Subtraction:", a - b)
print("Multiplication:", a * b)
print("Division:", a / b)

print("Add 5:", a + 5)
print("Multiply by 3:", a * 3)
print("Greater than 20:", a[a > 20])

marks_test = np.array([55, 82, 91, 67, 88])
result = marks_test[marks_test >= 80]
print("Marks >= 80:", result)


Addition: [11 22 33 44]
Subtraction: [ 9 18 27 36]
Multiplication: [ 10  40  90 160]
Division: [10. 10. 10. 10.]
Add 5: [15 25 35 45]
Multiply by 3: [ 30  60  90 120]
Greater than 20: [30 40]
Marks >= 80: [82 91 88]


### What Vectorized Operations Do:
Vectorized operations execute batch calculations on entire arrays simultaneously without explicit Python `for` loops. NumPy delegates the computations directly to compiled low-level routines, resulting in 10x to 100x performance gains over vanilla Python iterations.



---
## Part F – Debugging Practice
In this section, we analyze four common NumPy bugs, identify their root causes, and provide corrected code.



In [6]:
# Problem 1: IndexError with Out-of-Bounds Index
# Bug: arr = np.array([10, 20, 30]); print(arr[3]) -> IndexError
arr_p1 = np.array([10, 20, 30])
print("[Problem 1 Fixed] Last element at valid index 2:", arr_p1[2])

# Problem 2: Slicing Beyond Boundaries
# Note: arr[1:10] does NOT raise IndexError; slicing gracefully clips to array bounds.
arr_p2 = np.array([1, 2, 3, 4])
print("[Problem 2 Idiomatic] Slicing to end with arr[1:]:", arr_p2[1:])

# Problem 3: Shape Incompatibility in Element-wise Arithmetic
# Bug: a = np.array([1, 2, 3]); b = np.array([10, 20]); print(a + b) -> ValueError
a_p3 = np.array([1, 2, 3])
b_p3 = np.array([10, 20, 30])
print("[Problem 3 Fixed] Compatible element-wise sum:", a_p3 + b_p3)

# Problem 4: Multi-dimensional Out-of-Bounds Row Index
# Bug: arr = np.array([[1, 2], [3, 4]]); print(arr[2, 0]) -> IndexError (only rows 0 and 1 exist)
arr_p4 = np.array([[1, 2], [3, 4]])
print("[Problem 4 Fixed] Row 1, Col 0 element:", arr_p4[1, 0])


[Problem 1 Fixed] Last element at valid index 2: 30
[Problem 2 Idiomatic] Slicing to end with arr[1:]: [2 3 4]
[Problem 3 Fixed] Compatible element-wise sum: [11 22 33]
[Problem 4 Fixed] Row 1, Col 0 element: 3


### Debugging Summary Table

| Problem | Original Buggy Code | Error Raised / Behavior | Root Cause | Correction |
| :--- | :--- | :--- | :--- | :--- |
| **1** | `arr[3]` on size 3 | `IndexError: index 3 is out of bounds` | Zero-based indexing means valid indices are 0, 1, 2. | `arr[2]` or `arr[-1]` |
| **2** | `arr[1:10]` on size 4 | No error (returns `[2, 3, 4]`) | Python slicing clamps indices to range without failing. | `arr[1:]` (cleaner) |
| **3** | `a + b` (shapes `(3,)` & `(2,)`) | `ValueError: operands could not be broadcast` | Incompatible shapes cannot be paired element-wise. | Ensure matching dimensions |
| **4** | `arr[2, 0]` on `(2, 2)` matrix | `IndexError: index 2 is out of bounds for axis 0` | Axis 0 has only 2 rows (indices 0 and 1). | `arr[1, 0]` |



---
## Part G – Main Hands-on Activity: Student Marks Analyzer
A practical data analysis scenario analyzing examination scores across 4 students and 3 subjects using NumPy.



In [7]:
# Marks Matrix: Rows = Students (4), Columns = Subjects (3)
marks = np.array([
    [78, 85, 92],
    [66, 74, 81],
    [90, 88, 95],
    [55, 69, 72]
])

print("Marks Array:\n", marks)
print("Dimensions:", marks.ndim)
print("Shape:", marks.shape)
print("First Student Marks:", marks[0])
print("Second Subject Marks:", marks[:, 1])

# Vectorized bonus (+5) capped at 100
adjusted_marks = np.minimum(marks + 5, 100)
print("\nAdjusted Marks:\n", adjusted_marks)

# Student averages across subjects (axis=1)
averages = np.mean(adjusted_marks, axis=1)
print("Student Averages:", averages)

# Extremum analysis
print("Highest Mark (Original):", np.max(marks))
print("Adjusted Scores >= 80:", adjusted_marks[adjusted_marks >= 80])
print("Highest Adjusted Mark:", np.max(adjusted_marks))


Marks Array:
 [[78 85 92]
 [66 74 81]
 [90 88 95]
 [55 69 72]]
Dimensions: 2
Shape: (4, 3)
First Student Marks: [78 85 92]
Second Subject Marks: [85 74 88 69]

Adjusted Marks:
 [[ 83  90  97]
 [ 71  79  86]
 [ 95  93 100]
 [ 60  74  77]]
Student Averages: [90.         78.66666667 96.         70.33333333]
Highest Mark (Original): 95
Adjusted Scores >= 80: [ 83  90  97  86  95  93 100]
Highest Adjusted Mark: 100


### Student Marks Analysis Summary

| Student ID | Subject 1 (Adj) | Subject 2 (Adj) | Subject 3 (Adj) | Average Score | Performance Classification |
| :--- | :---: | :---: | :---: | :---: | :--- |
| **Student 1** | 83 | 90 | 97 | **90.00** | Distinction (Grade A) |
| **Student 2** | 71 | 79 | 86 | **78.67** | First Class (Grade B) |
| **Student 3** | 95 | 93 | 100 | **96.00** | Outstanding (Grade A+) |
| **Student 4** | 60 | 74 | 77 | **70.33** | Pass (Grade C) |



---
## Part H – Conceptual Reflection: Explain in Your Own Words

### 1. What is NumPy, and why is it useful?
**Answer:** NumPy (*Numerical Python*) is the fundamental Python package for scientific computing. It provides high-performance multidimensional array objects (`ndarray`) and vectorized mathematical routines implemented in C/Fortran. It is essential because it speeds up numerical operations dramatically compared to standard Python loops while providing concise, readable syntax.

### 2. What is a NumPy array?
**Answer:** A NumPy array is a contiguous, fixed-size, homogeneous memory block holding data elements of the same type organized across one or more dimensions.

### 3. What is the difference between a Python list and a NumPy array?
**Answer:** 
- A **Python list** is an array of object pointers, can hold heterogeneous data types, incurs dynamic type checking overhead, and requires explicit iterative loops for math.
- A **NumPy array** stores unboxed, homogeneous values contiguously in memory, consumes significantly less RAM, and executes operations directly at hardware speeds via vectorization.

### 4. What do ndim, shape, size, and dtype represent?
**Answer:**
- `ndim`: The number of axes / dimensions (e.g., 1 for 1D, 2 for 2D).
- `shape`: A tuple indicating the extent/length along each axis.
- `size`: The total count of all elements in the entire array (`product of shape`).
- `dtype`: The data type of the array elements (e.g., `int32`, `float64`).

### 5. What is indexing? What is negative indexing?
**Answer:**
- **Indexing** accesses an individual item at a specific location using 0-based integer coordinates (`arr[0]`).
- **Negative indexing** accesses elements counting from the end of the array backwards (`arr[-1]` for last, `arr[-2]` for second-to-last).

### 6. What is slicing?
**Answer:** Slicing is the technique of selecting a subarray or sequence of elements using the syntax `[start:stop:step]` without copying data whenever possible.

### 7. What is a vectorized operation?
**Answer:** A vectorized operation applies a mathematical or comparison operation to all array elements concurrently using pre-compiled low-level loops, avoiding explicit Python `for` loops.

### 8. Why must arrays generally have compatible shapes for element-wise arithmetic?
**Answer:** For element-wise operations to evaluate unambiguously, every element in one array must map to a corresponding element in the other according to NumPy broadcasting rules (dimensions must either match or one of them must be 1).



---
## Student Submission Checklist
- [x] I can create NumPy arrays.
- [x] I can identify 1D, 2D, and 3D arrays.
- [x] I understand `ndim`, `shape`, `size`, and `dtype`.
- [x] I can access elements using indexing.
- [x] I can extract data using slicing.
- [x] I can perform arithmetic without element-by-element loops.
- [x] I can filter array values using Boolean conditions.
- [x] I tested and corrected the debugging problems.
- [x] I can explain the main concepts in my own words.

